src/ccache.c

```diff
@@ -0,0 +1,523 @@
+	    /* ccache = */ true);
+
+	bin->ccache_bin_entry[nflush - 1] = ptr;
+	__asm__ __volatile__ ("" ::: "memory" );
+	bin->head = &bin->ccache_bin_entry[nflush - 1];
+
+	/* TODO: re-consider return true or false on success. This caused a bug
+	 * becuase the return value was forwarded by ccache_free, which should
+	 * return true on success, but this was returning 'false'
+	 */
+	return true;
+}
+
+static void
+ccache_bin_full_flush_unsafe(tsd_t *tsd, ccache_bin_t *bin, szind_t binind,
+    bool small) {
+	const unsigned nflush = ccache_bin_ncached_elements(bin);
+
+	CACHE_BIN_PTR_ARRAY_DECLARE(ptrs, nflush);
+	ptrs.ptr = &bin->ccache_bin_entry[CCACHE_BIN_ELEMENTS - nflush];
+
+	arena_flush(tsd, *tsd_arenap_get(tsd), &ptrs, binind, nflush, NULL,
+	    small, true);
+
+	bin->head = (void *)&bin->head;
+}
+
+static void
+ccache_bump_tstats(tsd_t *tsd, szind_t binind) {
+	cache_bin_stats_t *stats = ccache_bin_tstats_get(tsd, binind);
+	++stats->nrequests;
+}
+
+/* Depends on booted tcache */
+bool
+ccache_init(tsdn_t *tsdn, base_t *base) {
+	assert(ncpus > 0);
+
+	ccache_minind = sz_size2index(tcache_maxclass) + 1;
+	ccache_maxind = ccache_minind + ccached_nclasses - 1;
+	ccache_maxclass = sz_index2size(ccache_maxind);
+	ccache_bin_size = PAGE;
+	ccache_size = sizeof(ccache_t) * ncpus;
+	ccache = (ccache_t*)base_alloc(tsdn, base, ccache_size,
+	    PAGE);
+	if (ccache == NULL) {
+		return true;
+	}
+	for (unsigned i = 0; i < ncpus; ++i) {
+		ccache_t *ith_cache = &ccache[i];
+		for (unsigned j = 0; j < ccached_nclasses; ++j) {
+			ccache_bin_t *bin = &ith_cache->bins[j];
+			bin->head = (void **)&bin->head;
+		}
+	}
+
+	return false;
+}
+
+void*
+ccache_alloc(tsd_t *tsd, arena_t *arena, size_t size, szind_t ind,
+    bool zero, bool small) {
+	/*
+	 * Ccache is disabled for manual arenas due to difficulties handling
+	 * arena reset.
+	 */
+	assert(arena_is_auto(arena));
+
+	void *ret = NULL;
+	volatile rseq_t *rseq_abi = &tsd_ccache_tdatap_get(tsd)->rseq_abi;
+	int cpu;
+	ccache_t *ccache;
+	ccache_bin_t *bin;
+restart:
+	cpu = rseq_abi->cpu_id;
+	ccache = ccache_get(cpu);
+	bin = ccache_bin_get(ccache, ind);
+	__asm__ __volatile__ goto(
+	    /* aw == Allocatable and writable section */
+	    ".pushsection __rseq_table, \"aw\"\n\t"
+	    ".balign 32\n\t"
+	    "cs_obj:\n\t"
+	    /* rseq_cs.version rseq_cs.flags */
+	    ".long 0, 0\n\t"
+	    /* rseq_cs.start_ip rseq_cs.post_commit_offset rseq_cs.abort_ip */
+	    ".quad rseq_cs_start%=, rseq_cs_postcommit%= - rseq_cs_start%=, rseq_cs_abort%=\n\t"
+	    ".popsection\n\t"
+
+	    "leaq cs_obj(%%rip), %%rax\n\t"
+	    "movq %%rax, %[rseq_cs]\n\t"
+
+	    /* -------------------- Begin ------------------- */
+	    "rseq_cs_start%=:\n\t"
+	    "cmpl %[cpu_id], %[initial_cpu_id]\n\t"
+	    /* Restart in case of CPU id missmatch*/
+	    "jne %l[restart]\n\t"
+
+	    /* rax = head */
+	    "movq %c[head_offset](%[bin]), %%rax\n\t"
+	    /* If in fallback mode, then jump to fallback label */
```

Comment:
Do we use head==null as a flag for the fallback mode? If so let's document when / how that happens.

Reply:
Yes, `head==null` means fallback mode. I'll document the state machine once it's finalized.